bee-protocol/src/types/key_manager.rs

```diff
@@ -0,0 +1,82 @@
+// Copyright 2020 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+//! A module that provides a milestone key range manager type.
+
+use crate::types::key_range::MilestoneKeyRange;
+
+use bee_message::milestone::MilestoneIndex;
+
+use std::collections::HashSet;
+
+/// A key manager is managing a set of `MilestoneKeyRange`s.
+#[derive(Clone)]
+pub struct MilestoneKeyManager {
+    min_threshold: usize,
+    key_ranges: Box<[MilestoneKeyRange]>,
+}
+
+impl MilestoneKeyManager {
+    /// Creates a new `MilestoneKeyManager`.
+    pub fn new(min_threshold: usize, mut key_ranges: Box<[MilestoneKeyRange]>) -> Self {
+        key_ranges.sort();
+
+        Self {
+            min_threshold,
+            key_ranges,
+        }
+    }
+
+    /// Returns the minimum threshold of the `MilestoneKeyManager`.
+    pub fn min_threshold(&self) -> usize {
+        self.min_threshold
+    }
+
+    /// Returns a set of public keys applicable for a given milestone index.
+    pub fn get_public_keys(&self, index: MilestoneIndex) -> HashSet<String> {
+        let mut public_keys = HashSet::with_capacity(self.key_ranges.len());
+
+        for key_range in self.key_ranges.iter() {
+            if key_range.start() <= index {
+                if key_range.end() >= index
+                // start == end means the key is valid forever.
+                || key_range.start() == key_range.end()
+                {
+                    public_keys.insert(key_range.public_key().clone());
+                }
+                continue;
+            }
+            break;
+        }
+
+        public_keys
+    }
+}
+
+#[test]
+fn key_manager_is_sorted() {
```

Comment:
No `tests` module with `#[cfg(test)]` to avoid compiling outside of tests?

Reply:
Mistake, thanks